File: csharp/Shield/GroupEncryption.cs
using System;
using System.Collections.Generic;
using System.Security.Cryptography;

namespace Guard8.Shield
{
    /// <summary>
    /// GroupEncryption - Multi-recipient encryption.
    ///
    /// Encrypt once for multiple recipients, each can decrypt with their own key.
    /// Uses a group key for message encryption, then encrypts the group key
    /// separately for each member.
    /// </summary>
    public class GroupEncryption : IDisposable
    {
        private const int NonceSize = 16;
        private const int MacSize = 16;

        private byte[] _groupKey;
        private readonly Dictionary<string, byte[]> _members = new();
        private bool _disposed;

        /// <summary>
        /// Create group encryption with generated group key.
        /// </summary>
        public GroupEncryption()
        {
            _groupKey = Shield.RandomBytes(32);
        }

        /// <summary>
        /// Create group encryption with specified group key.
        /// </summary>
        public GroupEncryption(byte[] groupKey)
        {
            if (groupKey.Length != 32)
                throw new ArgumentException("Group key must be 32 bytes", nameof(groupKey));

            _groupKey = new byte[32];
            Array.Copy(groupKey, _groupKey, 32);
        }

        /// <summary>
        /// Add a member to the group.
        /// </summary>
        public void AddMember(string memberId, byte[] sharedKey)
        {
            if (sharedKey.Length != 32)
                throw new ArgumentException("Shared key must be 32 bytes", nameof(sharedKey));

            var keyCopy = new byte[32];
            Array.Copy(sharedKey, keyCopy, 32);
            _members[memberId] = keyCopy;
        }

        /// <summary>
        /// Remove a member from the group.
        /// </summary>
        public bool RemoveMember(string memberId)
        {
            return _members.Remove(memberId);
        }

        /// <summary>
        /// Get list of member IDs.
        /// </summary>
        public IEnumerable<string> Members => _members.Keys;

        /// <summary>
        /// Encrypt for all group members.
        /// </summary>
        public Dictionary<string, object> Encrypt(byte[] plaintext)
        {
            // Encrypt message with group key
            byte[] ciphertext = EncryptBlock(_groupKey, plaintext);

            // Encrypt group key for each member
            var encryptedKeys = new Dictionary<string, string>();
            foreach (var (memberId, memberKey) in _members)
            {
                byte[] encKey = EncryptBlock(memberKey, _groupKey);
                encryptedKeys[memberId] = Convert.ToBase64String(encKey);
            }

            return new Dictionary<string, object>
            {
                ["version"] = 1,
                ["ciphertext"] = Convert.ToBase64String(ciphertext),
                ["keys"] = encryptedKeys
            };
        }

        /// <summary>
        /// Decrypt as a group member.
        /// </summary>
        public static byte[] Decrypt(Dictionary<string, object> encrypted, string memberId, byte[] memberKey)
        {
            if (!encrypted.TryGetValue("keys", out var keysObj) ||
                keysObj is not Dictionary<string, string> keys ||
                !keys.TryGetValue(memberId, out var encryptedKeyB64))
            {
                return null;
            }

            // Decrypt group key
            byte[] encryptedGroupKey = Convert.FromBase64String(encryptedKeyB64);
            byte[] groupKey = DecryptBlock(memberKey, encryptedGroupKey);
            if (groupKey == null)
                return null;

            // Decrypt message
            byte[] ciphertext = Convert.FromBase64String((string)encrypted["ciphertext"]);
            return DecryptBlock(groupKey, ciphertext);
        }

        /// <summary>
        /// Rotate the group key.
        /// </summary>
        public byte[] RotateKey()
        {
            byte[] oldKey = _groupKey;
            _groupKey = Shield.RandomBytes(32);
            return oldKey;
        }

        // ============== Helper Methods ==============

        private static byte[] EncryptBlock(byte[] key, byte[] data)
        {
            byte[] nonce = Shield.RandomBytes(NonceSize);
            byte[] keystream = GenerateKeystream(key, nonce, data.Length);
            byte[] ciphertext = new byte[data.Length];
            for (int i = 0; i < data.Length; i++)
                ciphertext[i] = (byte)(data[i] ^ keystream[i]);

            byte[] macData = new byte[NonceSize + ciphertext.Length];
            Array.Copy(nonce, 0, macData, 0, NonceSize);
            Array.Copy(ciphertext, 0, macData, NonceSize, ciphertext.Length);
            byte[] mac = Shield.HmacSha256(key, macData);

            byte[] result = new byte[NonceSize + ciphertext.Length + MacSize];
            Array.Copy(nonce, 0, result, 0, NonceSize);
            Array.Copy(ciphertext, 0, result, NonceSize, ciphertext.Length);
            Array.Copy(mac, 0, result, NonceSize + ciphertext.Length, MacSize);

            return result;
        }

        private static byte[] DecryptBlock(byte[] key, byte[] encrypted)
        {
            if (encrypted.Length < NonceSize + MacSize)
                return null;

            byte[] nonce = new byte[NonceSize];
            Array.Copy(encrypted, 0, nonce, 0, NonceSize);

            int ciphertextLen = encrypted.Length - NonceSize - MacSize;
            byte[] ciphertext = new byte[ciphertextLen];
            Array.Copy(encrypted, NonceSize, ciphertext, 0, ciphertextLen);

            byte[] receivedMac = new byte[MacSize];
            Array.Copy(encrypted, encrypted.Length - MacSize, receivedMac, 0, MacSize);

            byte[] macData = new byte[NonceSize + ciphertextLen];
            Array.Copy(nonce, 0, macData, 0, NonceSize);
            Array.Copy(ciphertext, 0, macData, NonceSize, ciphertextLen);
            byte[] expectedMac = Shield.HmacSha256(key, macData);

            if (!Shield.ConstantTimeEquals(receivedMac, expectedMac, MacSize))
                return null;

            byte[] keystream = GenerateKeystream(key, nonce, ciphertextLen);
            byte[] decrypted = new byte[ciphertextLen];
            for (int i = 0; i < ciphertextLen; i++)
                decrypted[i] = (byte)(ciphertext[i] ^ keystream[i]);

            return decrypted;
        }

        private static byte[] GenerateKeystream(byte[] key, byte[] nonce, int length)
        {
            int numBlocks = (length + 31) / 32;
            byte[] keystream = new byte[numBlocks * 32];

            for (int i = 0; i < numBlocks; i++)
            {
                byte[] block = new byte[32 + NonceSize + 4];
                Array.Copy(key, 0, block, 0, 32);
                Array.Copy(nonce, 0, block, 32, NonceSize);
                BitConverter.GetBytes(i).CopyTo(block, 32 + NonceSize);

                byte[] hash = Shield.Sha256(block);
                Array.Copy(hash, 0, keystream, i * 32, 32);
            }

            byte[] result = new byte[length];
            Array.Copy(keystream, result, length);
            return result;
        }

        public void Dispose()
        {
            if (!_disposed)
            {
                Shield.SecureWipe(_groupKey);
                foreach (var key in _members.Values)
                    Shield.SecureWipe(key);
                _disposed = true;
            }
        }
    }
}

File: swift/Sources/Shield/GroupEncryption.swift
import Foundation
import CommonCrypto

/// GroupEncryption - Multi-recipient encryption.
///
/// Encrypt once for multiple recipients, each can decrypt with their own key.
/// Uses a group key for message encryption, then encrypts the group key
/// separately for each member.
public class GroupEncryption {
    private static let nonceSize = 16
    private static let macSize = 16

    private var groupKey: [UInt8]
    private var members: [String: [UInt8]] = [:]

    /// Create group encryption with generated group key.
    public init() {
        self.groupKey = Shield.randomBytes(32) ?? [UInt8](repeating: 0, count: 32)
    }

    /// Create group encryption with specified group key.
    public init(groupKey: [UInt8]) throws {
        guard groupKey.count == 32 else {
            throw ShieldError.invalidKeySize
        }
        self.groupKey = groupKey
    }

    /// Add a member to the group.
    public func addMember(_ memberId: String, sharedKey: [UInt8]) throws {
        guard sharedKey.count == 32 else {
            throw ShieldError.invalidKeySize
        }
        members[memberId] = sharedKey
    }

    /// Remove a member from the group.
    public func removeMember(_ memberId: String) -> Bool {
        return members.removeValue(forKey: memberId) != nil
    }

    /// Get list of member IDs.
    public var memberIds: [String] {
        return Array(members.keys)
    }

    /// Encrypt for all group members.
    public func encrypt(_ plaintext: [UInt8]) throws -> [String: Any] {
        // Encrypt message with group key
        let ciphertext = try encryptBlock(key: groupKey, data: plaintext)

        // Encrypt group key for each member
        var encryptedKeys: [String: String] = [:]
        for (memberId, memberKey) in members {
            let encKey = try encryptBlock(key: memberKey, data: groupKey)
            encryptedKeys[memberId] = Data(encKey).base64EncodedString()
        }

        return [
            "version": 1,
            "ciphertext": Data(ciphertext).base64EncodedString(),
            "keys": encryptedKeys
        ]
    }

    /// Decrypt as a group member.
    public static func decrypt(_ encrypted: [String: Any], memberId: String, memberKey: [UInt8]) -> [UInt8]? {
        guard let keys = encrypted["keys"] as? [String: String],
              let encryptedKeyB64 = keys[memberId],
              let encryptedKeyData = Data(base64Encoded: encryptedKeyB64) else {
            return nil
        }

        // Decrypt group key
        guard let groupKey = try? decryptBlock(key: memberKey, encrypted: Array(encryptedKeyData)) else {
            return nil
        }

        // Decrypt message
        guard let ciphertextB64 = encrypted["ciphertext"] as? String,
              let ciphertextData = Data(base64Encoded: ciphertextB64) else {
            return nil
        }

        return try? decryptBlock(key: groupKey, encrypted: Array(ciphertextData))
    }

    /// Rotate the group key.
    public func rotateKey() -> [UInt8] {
        let oldKey = groupKey
        groupKey = Shield.randomBytes(32) ?? [UInt8](repeating: 0, count: 32)
        return oldKey
    }

    /// Wipe keys from memory.
    public func wipe() {
        Shield.secureWipe(&groupKey)
        for key in members.keys {
            if var memberKey = members[key] {
                Shield.secureWipe(&memberKey)
            }
        }
        members.removeAll()
    }

    // MARK: - Helper Methods

    private func encryptBlock(key: [UInt8], data: [UInt8]) throws -> [UInt8] {
        guard let nonce = Shield.randomBytes(GroupEncryption.nonceSize) else {
            throw ShieldError.randomGenerationFailed
        }

        let keystream = generateKeystream(key: key, nonce: nonce, length: data.count)
        var ciphertext = [UInt8](repeating: 0, count: data.count)
        for i in 0..<data.count {
            ciphertext[i] = data[i] ^ keystream[i]
        }

        let macData = nonce + ciphertext
        let mac = Shield.hmacSha256(key: key, data: macData)

        return nonce + ciphertext + Array(mac.prefix(GroupEncryption.macSize))
    }

    private static func decryptBlock(key: [UInt8], encrypted: [UInt8]) throws -> [UInt8] {
        guard encrypted.count >= nonceSize + macSize else {
            throw ShieldError.ciphertextTooShort
        }

        let nonce = Array(encrypted.prefix(nonceSize))
        let ciphertext = Array(encrypted[nonceSize..<(encrypted.count - macSize)])
        let receivedMac = Array(encrypted.suffix(macSize))

        let macData = nonce + ciphertext
        let expectedMac = Array(Shield.hmacSha256(key: key, data: macData).prefix(macSize))

        guard Shield.constantTimeEquals(receivedMac, expectedMac) else {
            throw ShieldError.authenticationFailed
        }

        let keystream = generateKeystream(key: key, nonce: nonce, length: ciphertext.count)
        var decrypted = [UInt8](repeating: 0, count: ciphertext.count)
        for i in 0..<ciphertext.count {
            decrypted[i] = ciphertext[i] ^ keystream[i]
        }

        return decrypted
    }

    private static func generateKeystream(key: [UInt8], nonce: [UInt8], length: Int) -> [UInt8] {
        let numBlocks = (length + 31) / 32
        var keystream = [UInt8]()
        keystream.reserveCapacity(numBlocks * 32)

        for i in 0..<numBlocks {
            var block = key + nonce
            block.append(contentsOf: withUnsafeBytes(of: UInt32(i).littleEndian) { Array($0) })
            let hash = Shield.sha256(block)
            keystream.append(contentsOf: hash)
        }

        return Array(keystream.prefix(length))
    }
}
